File: node/transform.js
import fs from 'fs';
import { URL } from 'url';
import cleaners from './cleaners.js';
import MarkdownIt from 'markdown-it';
import path from 'path';
import configs from './firecrawl_to_trieve_config.js';

const MOCK_MODE = false;

// Use CONFIGS from firecrawl_to_trieve_config.js
const CONFIGS = configs.CONFIGS;

const markdown = new MarkdownIt();
const chunks = [];

function getTags(url) {
  const parsedUrl = new URL(url);
  const pathParts = parsedUrl.pathname.split('/');
  if (pathParts.includes('docs')) {
    const docsIndex = pathParts.indexOf('docs');
    // Return all non-empty tags after 'docs' except the last one
    return pathParts.slice(docsIndex + 1, -1).filter(tag => tag);
  }
  return [];
}

function getImages(markdownContent) {
  const imagePattern = /!\[.*?\]\((.*?\.(?:png|webp))\)/g;
  const matches = markdownContent.match(imagePattern);
  return matches ? matches.map(match => match.match(/\((.*?)\)/)[1]) : [];
}

function getTrackingId(url) {
  return url.replace('https://signoz.io/', '')
    .replace(/#/g, '-')
    .replace(/\s/g, '-')
    .replace(/:/g, '-')
    .replace(/\//g, '-')
    .replace(/--/g, '-')
    .trim()
    .replace(/^-|-$/g, '');
}

function getChunkHtml(content, pageTitle, headingText, startIndex, chunkEnd) {
  try {
      const chunkContent = chunkEnd !== null ?
          content.slice(startIndex, chunkEnd) :
          content.slice(startIndex);
      let chunkText = chunkContent.split('\n').join('\n').trim().replace(/^-+|-+$/g, '');
      let chunkHtml = chunkText.trim();

      chunkHtml = cleaners.cleanMultiColumnLinks(chunkHtml);
      chunkHtml = cleaners.cleanAnchortagHeadings(chunkHtml);
      chunkHtml = cleaners.cleanExtraNewlinesAfterLinks(chunkHtml);
      chunkHtml = cleaners.cleanDoubleAsteriskWhitespaceGaps(chunkHtml);
      chunkHtml = cleaners.cleanNewlineAndSpacesAfterLinks(chunkHtml);
      

      // Skip heading-only chunks
      if (chunkHtml.trim().split('\n').length <= 1) {
          return {"HEADING_ONLY": chunkHtml.trim()};
      }
  
      if (headingText === "") {
        chunkHtml = chunkHtml.trim().replace(/^-+|-+$/g, '');
      } else {
        const headingLine = `${pageTitle}: ${headingText}`;
        const lines = chunkHtml.split('\n');
        lines[0] = headingLine;
        chunkHtml = lines.join('\n');
      }
      return chunkHtml;
  } catch (e) {
      console.error(`Error processing chunk: ${e.message}`);
      throw e;
  }
}

function createChunkObject(chunkHtml, pageLink, headingLink, headingText, pageTagsSet, 
                     pageTitle, pageDescription) {
  const chunk = {
    chunk_html: chunkHtml,
    link: pageLink + headingLink,
    tags_set: pageTagsSet,
    image_urls: getImages(chunkHtml),
    tracking_id: getTrackingId(pageLink + headingLink),
    group_tracking_ids: [getTrackingId(pageLink)],
    timestamp: CONFIGS.TIMESTAMP,
    metadata: {
      title: pageTitle + (headingText.length > 0 ? ': ' + headingText : ''),
      page_title: pageTitle,
      page_description: pageDescription,
    }
  };

  if (CONFIGS.boost) {
    const boostPhrase = (pageTitle + " " + (headingText || '')).trim().replace(/^:\s+/, '');
    chunk.semantic_boost = {
      distance_factor: CONFIGS.semanticBoostDistanceFactor,
      phrase: boostPhrase
    };
    chunk.fulltext_boost = {
      boost_factor: CONFIGS.fulltextBoostFactor,
      phrase: boostPhrase
    };
  }

  return chunk;
}

export function headingMatches(content, pattern) {
  const matches = Array.from(content.matchAll(pattern));
  return matches.map(match => [
    match[2], // headingLink
    match[3], // headingText
    match[0]  // heading symbols and link
  ]);
}

export function splitContent(content, pattern) {
    const matches = headingMatches(content, pattern);
    const sections = [];

    let preSplitContent = content;
    if (matches.length > 0) {
        let firstMatch = matches[0];

        if (firstMatch) {
            preSplitContent = content.split(firstMatch[2])[0];
            if (preSplitContent.trim().length > 0) {
                sections.push(["", "", preSplitContent]);
            }
        }
    }

    for (let i = 0; i < matches.length; i++) {
        const match = matches[i];
        const [headingLink, headingText, matchString] = match;
        const start = content.indexOf(matchString);
        let end;

        if (i < matches.length - 1) {
            end = content.indexOf(matches[i + 1][2]);
        } else {
            end = content.length;
        }

        if (start === -1 || end === -1) {
            throw new Error('Invalid start or end index when splitting content');
        }

        sections.push([headingLink, headingText, content.slice(start, end)]);
    }

    return sections;
}

function processContent(pageMarkdown, pageTitle, pageLink, pageTagsSet,
                        pageDescription, maxWords = CONFIGS.maxWords, maxDepth = CONFIGS.maxDepth) {

  

  // Creates chunks from sections
  function createChunks(sections, currentTitle = '', depth = 0) {
    const localChunks = [];
    let isLastChunkHeadingOnly = false;

    sections.forEach(([headingLink, headingText, sectionContent]) => {
      if (isLastChunkHeadingOnly) {
        headingText = `${isLastChunkHeadingOnly} - ${headingText}`;
        isLastChunkHeadingOnly = false;
      }
      
      const chunkHtml = getChunkHtml(sectionContent, pageTitle, headingText, 0, null);
      

      try {
          isLastChunkHeadingOnly = chunkHtml["HEADING_ONLY"]
      } catch (TypeError) {
          isLastChunkHeadingOnly = false
      }
      if (isLastChunkHeadingOnly) {
        return; // Skip to the next iteration
      } else {

        


        const fullTitle = `${currentTitle}: ${headingText}`.replace(/^:\s+/, '');
        const chunkWordCount = chunkHtml.split(/\s+/).length;
        const isWithinChunkingConstrains = chunkWordCount <= maxWords || depth >= maxDepth;
        // true if the chunk is within the word limit or we've reached max depth.
        // we don't split further at max depth, even if over word limit

        if (isWithinChunkingConstrains) {
          // Create chunk if within word limit or max depth reached
          let chunk = createChunkObject(chunkHtml, pageLink, headingLink, headingText,
                                            pageTagsSet, pageTitle, pageDescription);
          chunk.metadata.title = fullTitle;
          localChunks.push(chunk);
        } else {
          // Try to split into subsections
          const subsections = splitContent(sectionContent, 
            /(\\n###+ \\[\\]\\((#.*?)\\))\\n(.*?)\\n/g);
          // regex to find subsections of the current section
          if (subsections.length > 0) {
            localChunks.push(...createChunks(subsections, fullTitle, depth + 1));
          } else {
            // If no subsections, force create a chunk
            let chunk = createChunkObject(chunkHtml, pageLink, headingLink, headingText,
                                                pageTagsSet, pageTitle, pageDescription);
            chunk.metadata.title = fullTitle;
            localChunks.push(chunk);
          }
        }
      }
    });

    return localChunks;
  }

  const topSections = splitContent(pageMarkdown, /(\n\[\]\((#.*?)\))\n(.*?)\n/g);
  return createChunks(topSections, pageTitle);
}

async function main() {
  // Choose dataset
  const chosenDataset = await configs.chooseDataset();
  const configuration = configs.getConfiguration(chosenDataset);

  // Get the latest crawl results file
  const { file: crawlResultsFile, timestamp: TIMESTAMP } = configs.getLatestCrawlFileAndTimestamp();

  if (!crawlResultsFile) {
    console.error('No crawl results file found');
    process.exit(1);
  }

  // Load the crawl results
  const crawlResults = JSON.parse(fs.readFileSync(crawlResultsFile, 'utf8'));
  console.log(`Loaded ${crawlResults.length} crawl results from ${crawlResultsFile}`);
  // Iterate through each item in the crawl results
  crawlResults.forEach(item => {
    const url = item.metadata.ogUrl;

    // Skip pages with pageStatusCode != 200
    try {
      if (item.metadata.pageStatusCode !== 200) return;
    } catch (error) {
      throw new Error(`pageStatusCode not found for url: ${url}`);
    }

    const pageLink = url;
    const pageTitle = item.metadata.ogTitle;
    const pageDescription = item.metadata.description || '';
    let pageMarkdown = item.markdown;
    const pageTagsSet = getTags(url);

    // Remove end matter
    pageMarkdown = cleaners.removeEndMatter(pageMarkdown);

    // If the page is less than 500 words, then make one chunk for the page (baseline)
    if (pageMarkdown.split(/\s+/).length < 500) {
      chunks.push(createChunkObject(getChunkHtml(pageMarkdown, pageTitle, '', 0, null), pageLink, '', '', pageTagsSet, 
                              pageTitle, pageDescription));
    } else {
      // Otherwise, create subpage chunks 
      chunks.push(...processContent(pageMarkdown, pageTitle, pageLink,
                                    pageTagsSet, pageDescription));
    }
  });

  // strip ":" from title if it ends with it
  chunks.forEach(chunk => {
    chunk.metadata.title = chunk.metadata.title.replace(/:\s*$/, '');
  });

  // render markdown to html
  chunks.forEach(chunk => {
    chunk.chunk_html = markdown.render(chunk.chunk_html);
    // Replace <br> with <br />
    chunk.chunk_html = chunk.chunk_html.replace(/<br>/g, '<br />');
    // Replace <hr> with <hr />
    chunk.chunk_html = chunk.chunk_html.replace(/<hr>/g, '<hr />');
    // Replace <img ...> with <img ... />
    chunk.chunk_html = chunk.chunk_html.replace(/<img\s+([^>]*)>/g, '<img $1 />');
  });

  if (MOCK_MODE) {
    console.log(`MOCK: Would generate ${chunks.length} chunks from ${crawlResultsFile}, saving to .json and .md files`);
    return;
  }

  // Save the chunks data to chunks.json
  const chunkFilename = `chunks_${TIMESTAMP}_${configuration.trieveDatasetName}.json`;
  fs.writeFileSync(chunkFilename, JSON.stringify(chunks, null, 2));

  console.log(`Saved ${chunks.length} chunks to ${chunkFilename}`);

  // Generate chunks.md
  const chunksMarkdown = chunks.map(chunk => 
    `link: ${chunk.link}\n` +
    `tag_set: ${chunk.tags_set.join(', ')}\n` +
    `image_urls: ${chunk.image_urls.join(', ')}\n` +
    `tracking_id: ${chunk.tracking_id}\n` +
    `group_tracking_ids: ${chunk.group_tracking_ids.join(', ')}\n` +
    `${chunk.chunk_html}\n` +
    '-'.repeat(80) + '\n\n'
  ).join('');

  fs.writeFileSync(`chunks_${configuration.trieveDatasetName}.md`, chunksMarkdown);

  console.log(`Generated chunks_${configuration.trieveDatasetName}.md with ${chunks.length} entries`);
}

main();

File: python/transform.py
import json
import re
from urllib.parse import urlparse
import cleaners
import markdown
import os

CONFIGS = {
    'boost': False,
    'max_words': 500,
    'max_depth': 3,
    'semantic_boost_distance_factor': 0.5,
    'fulltext_boost_factor': 5,
    'root_url': 'https://signoz.io/'
}

crawl_results_files = sorted(
    [f for f in os.listdir('.') if f.startswith('crawl_results') and f.endswith('.json')],
    reverse=True
)

crawl_results_file = crawl_results_files[0] if crawl_results_files else ''

if not crawl_results_file:
    print('No crawl results file found')
    exit(1)

# Load the crawl results
with open(crawl_results_file, 'r') as f:
    crawl_results = json.load(f)

# Get the timestamp from the crawl results file name
TIMESTAMP = '_'.join(crawl_results_file.split('_')[-2:]).split('.')[0]

chunks = []

def get_tags(url):
    parsed_url = urlparse(url)
    path_parts = parsed_url.path.split('/')
    if 'docs' in path_parts:
        docs_index = path_parts.index('docs')
        # Return all non-empty tags after 'docs' except the last one
        return [tag for tag in path_parts[docs_index + 1:-1] if tag]
    return []


def get_images(markdown):
    image_pattern = r'\!\[.*?\]\((.*?\.(?:png|webp))\)'
    return re.findall(image_pattern, markdown)


def get_tracking_id(url):
    # Strip leading elements:
    url = url.replace(CONFIGS['root_url'], '')
    # Replace multiple dashes with a single dash
    return url.replace('#', '-').replace(' ', '-').replace(':', '-').replace('/', '-').replace('--', '-').strip('-')


def get_chunk_html(content, page_title, headingtext, start_index, chunk_end):
    """
    Get the content for a sub-page
    """        
    try:
        if chunk_end is not None:
            chunk_content = content[start_index:chunk_end]
        else:
            chunk_content = content[start_index:]
        chunk_text = "\n".join(chunk_content.splitlines()).strip().strip('-')
        chunk_html = f"{chunk_text.strip()}"

    except ValueError as e:
        chunk_html = f"Error processing chunk: {str(e)}"
        raise e

    chunk_html = cleaners.clean_multi_column_links(chunk_html)
    chunk_html = cleaners.clean_double_newline_markdown_links(chunk_html)
    chunk_html = cleaners.clean_anchortag_headings(chunk_html)
    chunk_html = cleaners.clean_extra_newlines_after_links(chunk_html)
    chunk_html = cleaners.clean_double_asterisk_whitespace_gaps(chunk_html)
    
    
    
    # Skip heading-only chunks
    if len(chunk_html.strip().splitlines()) <= 1:
        return {"HEADING_ONLY": chunk_html.strip()}
    
    
    if headingtext == "":
        chunk_html = f"{chunk_html.strip().strip('-')}"
    else:
        heading_line = f"{page_title}: {headingtext}"
        chunk_html = chunk_html.replace(chunk_html.splitlines()[0], heading_line)

    return chunk_html


def create_chunk(chunk_html, page_link, headinglink, headingtext, page_tag_set, page_title, page_description):
    if page_title.endswith(f": {headingtext}"):
        title = page_title
    else: 
        title = page_title + (': ' + headingtext if headingtext else '')
    chunk = {
        'chunk_html': chunk_html,
        'link': page_link+headinglink,
        'tag_set': page_tag_set,
        'image_urls': get_images(chunk_html),
        'tracking_id': get_tracking_id(page_link+headinglink),
        'group_tracking_ids': [get_tracking_id(page_link)],
        'timestamp': TIMESTAMP,
        'metadata': {
            'title': title,
            'page_title': page_title,
            'page_description': page_description,
        }
    }
    if CONFIGS['boost']:
        boost_phrase = (page_title + " " + headingtext).strip(': ')
        chunk['semantic_boost'] = {
            "distance_factor": CONFIGS['semantic_boost_distance_factor'],
            "phrase": boost_phrase
        }
        chunk['fulltext_boost'] = {
            "boost_factor": CONFIGS['fulltext_boost_factor'],
            "phrase": boost_phrase
        }

    return chunk

def process_content(page_markdown, page_title, page_link, page_tag_set,
                    page_description, max_words=CONFIGS['max_words'], max_depth=CONFIGS['max_depth']):
    def split_content(content, pattern, headingtext, headinglink):
        matches = re.findall(pattern, content, re.DOTALL)

        sections = []
        if matches:
            pre_split_content = content.split(matches[0][0])
            if len(pre_split_content[0].strip()) > 0:
                # sections contain: headinglink, headingtext, section_content
                sections.append((headinglink, headingtext, pre_split_content[0]))
        for i, match in enumerate(matches):
            matchstring, headinglink, headingtext = match
            start = content.index(matchstring)
            end = len(content) if i == len(matches) - 1 else content.index(matches[i+1][0])
            sections.append((headinglink, headingtext, content[start:end]))
        return sections

    def create_chunks(sections, current_title='', depth=0):
        local_chunks = []
        last_chunk_heading_only = False
        
        for headinglink, headingtext, section_content in sections:
            if last_chunk_heading_only:
                headingtext = last_chunk_heading_only + " - " + headingtext
                last_chunk_heading_only = False

            full_title = (f"{current_title}: {headingtext}" if current_title != headingtext
                          else headingtext).strip(': ')
            chunk_html = get_chunk_html(section_content, page_title, headingtext, 0, None)
            try:
                last_chunk_heading_only = chunk_html["HEADING_ONLY"]
            except TypeError:
                last_chunk_heading_only = False
            if last_chunk_heading_only:
                continue
            
            if len(chunk_html.split()) <= max_words or depth >= max_depth:
                # Create chunk if within word limit or max depth reached
                chunk = create_chunk(chunk_html, page_link, headinglink, headingtext,
                                     page_tag_set, page_title, page_description)
                if full_title.endswith(f"{headingtext}: {headingtext}"):
                    full_title = full_title.replace(f": {headingtext}", "", 1)
                chunk['metadata']['title'] = full_title
                local_chunks.append(chunk)
            else:
                # Try to split into subsections
                subsections = split_content(section_content, r'(\n###+ \[\]\((#.*?)\))\n(.*?)\n', headingtext, headinglink)
                if subsections:
                    local_chunks.extend(create_chunks(subsections,
                                                      current_title=full_title,
                                                      depth=depth + 1))
                else:
                    # If no subsections, force create a chunk
                    chunk = create_chunk(chunk_html, page_link, headinglink, headingtext,
                                         page_tag_set, page_title, page_description)
                    local_chunks.append(chunk)
        return local_chunks

    # Start with top-level headings
    top_sections = split_content(page_markdown, r'(\n\[\]\((#.*?)\))\n(.*?)\n', "", "")
    return create_chunks(top_sections, current_title=page_title)

def main():

    # Iterate through each item in the crawl results
    for item in crawl_results:
        url = item['metadata']['ogUrl']

        # Skip pages with pageStatusCode != 200
        try:
            if item['metadata']['pageStatusCode'] != 200:
                continue
        except KeyError:
            raise KeyError(f"pageStatusCode not found for url: {url}")

        page_link = url
        page_title = item['metadata']['ogTitle']
        page_description = item['metadata'].get('description', '')
        page_markdown = item['markdown']
        page_tag_set = get_tags(url)

        # Remove end matter
        page_markdown = cleaners.remove_end_matter(page_markdown)

        # If the page is less than 500 words, then make one chunk for the page (baseline)
        if len(page_markdown.split(" ")) < 500:
            chunk_html = get_chunk_html(page_markdown, page_title, '', 0, None)
            chunk = create_chunk(chunk_html, page_link, '', '', page_tag_set, page_title, page_description)
            chunks.append(chunk)
            continue

        # Otherwise, create subpage chunks 
        chunks.extend(process_content(page_markdown, page_title, page_link,
                                    page_tag_set, page_description))


    # render markdown to html
    for chunk in chunks:
        chunk['chunk_html'] = markdown.markdown(chunk['chunk_html'])


    # Save the chunks data to chunks.json
    chunk_filename = f'chunks_{TIMESTAMP}.json'
    if CONFIGS['boost']:
        chunk_filename = f'chunks_{TIMESTAMP}_boost.json'
    with open(chunk_filename, 'w') as f:
        json.dump(chunks, f, indent=2)

    print(f"Saved {len(chunks)} chunks to {chunk_filename}")

    # Generate chunks.md
    with open('chunks.md', 'w') as f:
        for chunk in chunks:
            f.write(f"link: {chunk['link']}\n")
            f.write(f"tag_set: {', '.join(chunk['tag_set'])}\n")
            f.write(f"image_urls: {', '.join(chunk['image_urls'])}\n")
            f.write(f"tracking_id: {chunk['tracking_id']}\n")
            f.write(f"group_tracking_ids: {', '.join(chunk['group_tracking_ids'])}\n")
            f.write(chunk['chunk_html'])
            f.write("\n" + "-" *80 + "\n\n")

    print(f"Generated chunks.md with {len(chunks)} entries")

if __name__ == "__main__":
    main()